Bug in Null condition function?

According to [AWS document](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition_operators.html#Conditions_Null), it seems that value `true` means that the specified key should be NULL - i.e. nonexistent. But [the code](https://github.com/minio/minio/blob/21d8c0fd13cfb775c1c5ddd05e01feea07d81674/pkg/policy/condition/nullfunc.go#L39) is implemented the other way around.

I've created a simple test function as follows. Just add this code snippet at the end of [statement_test.go](https://github.com/minio/minio/blob/master/pkg/policy/statement_test.go) and run. I've used [RELEASE.2018-12-13T02-04-19Z](https://github.com/minio/minio/tree/RELEASE.2018-12-13T02-04-19Z) by the way.
```go
func TestStatement_NullFunc(t *testing.T) {
	bucket := "bkt"
	object := "object"
	action := Action(PutObjectAction)
	principal := "Q3AM3UQ867SPQQA43P2F"
	condKey := condition.Key(condition.S3XAmzServerSideEncryption)
	condVal := "AES256"

	f, _ := condition.NewNullFunc(
		condKey,
		true,
	)

	stmt := NewStatement(
		Deny,
		NewPrincipal(principal),
		NewActionSet(action),
		NewResourceSet(NewResource(bucket, object)),
		condition.NewFunctions(f),
	)
	pol, _ := json.MarshalIndent(&stmt, "", "  ")
	t.Logf("%s", pol)

	args := Args{
		AccountName: principal,
		Action:      action,
		BucketName:  bucket,
		ConditionValues: map[string][]string{
			condKey.Name(): {condVal},
		},
		ObjectName: object,
		IsOwner: false,
	}
	res := stmt.IsAllowed(args)
	expected := true
	if expected != res {
		t.Fatalf("expected: %v, got: %v", expected, res)
	}
}
```

The execution results are as follows:
```
=== RUN   TestStatement_NullFunc
--- FAIL: TestStatement_NullFunc (0.00s)
	statement_test.go:597: {
		  "Effect": "Deny",
		  "Principal": {
		    "AWS": [
		      "Q3AM3UQ867SPQQA43P2F"
		    ]
		  },
		  "Action": [
		    "s3:PutObject"
		  ],
		  "Resource": [
		    "arn:aws:s3:::bkt/object"
		  ],
		  "Condition": {
		    "Null": {
		      "s3:x-amz-server-side-encryption": [
		        true
		      ]
		    }
		  }
		}
	statement_test.go:612: expected: true, got: false
FAIL
```
The policy prescribes that a PutObject request will be DENIED if it does NOT have server-side encryption header `x-amz-server-side-encryption`. But test results say a request with that header present is not allowed. After a short debugging I'm no suspecting the following code snippet:
```go
pkg/policy/condition/nullfunc.go:

func (f nullFunc) evaluate(values map[string][]string) bool {
	requestValue := values[f.k.Name()]

	if f.value {
		return len(requestValue) != 0
	}

	return len(requestValue) == 0
}
```
In my opinion, the `if` condition should be negated, so that the evaluation results in true if the value is NOT exists. Negating it makes the test passing.

--- a/pkg/policy/condition/func_test.go
+++ b/pkg/policy/condition/func_test.go
@@ -53,12 +53,12 @@ func TestFunctionsEvaluate(t *testing.T) {
 		{case1Function, map[string][]string{
 			"x-amz-copy-source": {"mybucket/myobject"},
 			"SourceIp":          {"192.168.1.10"},
-		}, true},
+		}, false},
 		{case1Function, map[string][]string{
 			"x-amz-copy-source": {"mybucket/myobject"},
 			"SourceIp":          {"192.168.1.10"},
 			"Refer":             {"http://example.org/"},
-		}, true},
+		}, false},
 		{case1Function, map[string][]string{"x-amz-copy-source": {"mybucket/myobject"}}, false},
 		{case1Function, map[string][]string{"SourceIp": {"192.168.1.10"}}, false},
 		{case1Function, map[string][]string{
@@ -79,7 +79,7 @@ func TestFunctionsEvaluate(t *testing.T) {
 		result := testCase.functions.Evaluate(testCase.values)
 
 		if result != testCase.expectedResult {
-			t.Fatalf("case %v: expected: %v, got: %v\n", i+1, testCase.expectedResult, result)
+			t.Errorf("case %v: expected: %v, got: %v\n", i+1, testCase.expectedResult, result)
 		}
 	}
 }

--- a/pkg/policy/condition/ipaddressfunc.go
+++ b/pkg/policy/condition/ipaddressfunc.go
@@ -19,6 +19,7 @@ package condition
 import (
 	"fmt"
 	"net"
+	"net/http"
 	"sort"
 )
 
@@ -46,7 +47,12 @@ type ipAddressFunc struct {
 // falls in one of network or not.
 func (f ipAddressFunc) evaluate(values map[string][]string) bool {
 	IPs := []net.IP{}
-	for _, s := range values[f.k.Name()] {
+	requestValue, ok := values[http.CanonicalHeaderKey(f.k.Name())]
+	if !ok {
+		requestValue = values[f.k.Name()]
+	}
+
+	for _, s := range requestValue {
 		IP := net.ParseIP(s)
 		if IP == nil {
 			panic(fmt.Errorf("invalid IP address '%v'", s))

--- a/pkg/policy/condition/nullfunc.go
+++ b/pkg/policy/condition/nullfunc.go
@@ -18,17 +18,19 @@ package condition
 
 import (
 	"fmt"
+	"net/http"
 	"reflect"
 	"strconv"
 )
 
-// nullFunc - Null condition function. It checks whether Key is present in given
+// nullFunc - Null condition function. It checks whether Key is not present in given
 // values or not.
 // For example,
 //   1. if Key = S3XAmzCopySource and Value = true, at evaluate() it returns whether
-//      S3XAmzCopySource is in given value map or not.
-//   2. if Key = S3XAmzCopySource and Value = false, at evaluate() it returns whether
 //      S3XAmzCopySource is NOT in given value map or not.
+//   2. if Key = S3XAmzCopySource and Value = false, at evaluate() it returns whether
+//      S3XAmzCopySource is in given value map or not.
+// https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition_operators.html#Conditions_Null
 type nullFunc struct {
 	k     Key
 	value bool
@@ -37,13 +39,16 @@ type nullFunc struct {
 // evaluate() - evaluates to check whether Key is present in given values or not.
 // Depending on condition boolean value, this function returns true or false.
 func (f nullFunc) evaluate(values map[string][]string) bool {
-	requestValue := values[f.k.Name()]
-
-	if f.value {
-		return len(requestValue) != 0
+	requestValue, ok := values[http.CanonicalHeaderKey(f.k.Name())]
+	if !ok {
+		requestValue = values[f.k.Name()]
 	}
 
-	return len(requestValue) == 0
+	if f.value {
+		return len(requestValue) == 0
+	}
+
+	return len(requestValue) != 0
 }
 
 // key() - returns condition key which is used by this condition function.

--- a/pkg/policy/condition/nullfunc_test.go
+++ b/pkg/policy/condition/nullfunc_test.go
@@ -37,23 +37,23 @@ func TestNullFuncEvaluate(t *testing.T) {
 		values         map[string][]string
 		expectedResult bool
 	}{
-		{case1Function, map[string][]string{"prefix": {"true"}}, true},
-		{case1Function, map[string][]string{"prefix": {"false"}}, true},
-		{case1Function, map[string][]string{"prefix": {"mybucket/foo"}}, true},
-		{case1Function, map[string][]string{}, false},
-		{case1Function, map[string][]string{"delimiter": {"/"}}, false},
-		{case2Function, map[string][]string{"prefix": {"true"}}, false},
-		{case2Function, map[string][]string{"prefix": {"false"}}, false},
-		{case2Function, map[string][]string{"prefix": {"mybucket/foo"}}, false},
-		{case2Function, map[string][]string{}, true},
-		{case2Function, map[string][]string{"delimiter": {"/"}}, true},
+		{case1Function, map[string][]string{"prefix": {"true"}}, false},
+		{case1Function, map[string][]string{"prefix": {"false"}}, false},
+		{case1Function, map[string][]string{"prefix": {"mybucket/foo"}}, false},
+		{case1Function, map[string][]string{}, true},
+		{case1Function, map[string][]string{"delimiter": {"/"}}, true},
+		{case2Function, map[string][]string{"prefix": {"true"}}, true},
+		{case2Function, map[string][]string{"prefix": {"false"}}, true},
+		{case2Function, map[string][]string{"prefix": {"mybucket/foo"}}, true},
+		{case2Function, map[string][]string{}, false},
+		{case2Function, map[string][]string{"delimiter": {"/"}}, false},
 	}
 
 	for i, testCase := range testCases {
 		result := testCase.function.evaluate(testCase.values)
 
 		if result != testCase.expectedResult {
-			t.Fatalf("case %v: expected: %v, got: %v\n", i+1, testCase.expectedResult, result)
+			t.Errorf("case %v: expected: %v, got: %v\n", i+1, testCase.expectedResult, result)
 		}
 	}
 }

--- a/pkg/policy/condition/stringequalsfunc.go
+++ b/pkg/policy/condition/stringequalsfunc.go
@@ -18,6 +18,7 @@ package condition
 
 import (
 	"fmt"
+	"net/http"
 	"sort"
 	"strings"
 
@@ -44,7 +45,11 @@ type stringEqualsFunc struct {
 // evaluate() - evaluates to check whether value by Key in given values is in
 // condition values.
 func (f stringEqualsFunc) evaluate(values map[string][]string) bool {
-	requestValue := values[f.k.Name()]
+	requestValue, ok := values[http.CanonicalHeaderKey(f.k.Name())]
+	if !ok {
+		requestValue = values[f.k.Name()]
+	}
+
 	return !f.values.Intersection(set.CreateStringSet(requestValue...)).IsEmpty()
 }
 

--- a/pkg/policy/condition/stringlikefunc.go
+++ b/pkg/policy/condition/stringlikefunc.go
@@ -18,6 +18,7 @@ package condition
 
 import (
 	"fmt"
+	"net/http"
 	"sort"
 	"strings"
 
@@ -45,7 +46,12 @@ type stringLikeFunc struct {
 // evaluate() - evaluates to check whether value by Key in given values is wildcard
 // matching in condition values.
 func (f stringLikeFunc) evaluate(values map[string][]string) bool {
-	for _, v := range values[f.k.Name()] {
+	requestValue, ok := values[http.CanonicalHeaderKey(f.k.Name())]
+	if !ok {
+		requestValue = values[f.k.Name()]
+	}
+
+	for _, v := range requestValue {
 		if !f.values.FuncMatch(wildcard.Match, v).IsEmpty() {
 			return true
 		}

